### PR TITLE
Add docs for the instance customisation plugin API

### PR DESCRIPTION
**Pull request by dan-tw**
## Relevant components:
- [x] Docs

## Problem statement:
There were no docs explaining how to use the instance customisation plugin API

## Solution
Adds docs to the repo explaining how to build and deploy a instance customisation plugin

## Extra

To see how they will look in the actual repo, visit the fork/branch https://github.com/dan-tw/APIs/tree/docs

### CODE REVIEW

**Review by david-macpherson — APPROVED** [review]

File: docs/instance-customisation/deploy.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+## Deploying to the cloud
+
+Once you have built your instance customisation plugin it needs to be deployed to the cloud. The most common way is to deploy it to the same cluster that is running Scalable Pixel Streaming, however, you can deploy it anywhere that is publically accessible.
+
+### Building a container 
+
+In order to deploy your plugin to the Scalable Pixel Streaming cluster, it first needs to be put into a container. You can follow the docker guide [here](https://docs.docker.com/language/golang/build-images/) or check out the Dockerfile we have in examples [here](examples/golang/instance-customisation-plugin)
```
Comment (sarahkrivan) [inline docs/instance-customisation/deploy.md]: Change:
> put into a container

to: 

> containerised

and capitalise Docker in sentence 2

File: docs/instance-customisation/deploy.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+### Deploying to the cluster
+
+Scalable Pixel Streaming is installed onto a [kubernetes](https://kubernetes.io/) cluster. Depending on what cloud platform you've installed Scalable Pixel Streaming onto will depend on how you access your cluster. Refer to the [Scalable Pixel Streaming documentation](http://docs.beta.scalablestreaming.io/) for information on how to access your cluster.
```
Comment (sarahkrivan) [inline docs/instance-customisation/deploy.md]: Change:
>Depending on what cloud platform you've installed Scalable Pixel Streaming onto will depend on how you access your cluster.

to:

> The cloud platform you have installed Scalable Pixel Streaming onto will inform how you access your cluster.

File: docs/instance-customisation/overview.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+## What is Instance Customisation
+Scalable Pixel Streaming provides you with the option to configure and **deploy your own instance customisation plugins**, providing you with the ability to customise certain aspects of your Scalable Pixel Streaming application **as they are requested**. The benefit to deploying your own instance customisation plugin is that you can change your Scalable Pixel Streaming application at **runtime**.
+
+The instance customisation plugin API sends the ****Instance ID****, ****Player ID****, ****Plugin Options**** and ****Runtime Options**** as part of the `UpdateRuntimeOptionsRequest` request and responds with a version of the `RuntimeOptions`.
+
+### What is the difference between Application and Instance customisation?
+Applications can be configured with much of the same options that are available through an instance customisation plugin. The difference is that an Application’s options are applied to **every instance of that Application** and an instance customisation plugin can **allow instance configuration on a per user level** and even **allow an end user to influence your instance configuration**.
```
Comment (sarahkrivan) [inline docs/instance-customisation/overview.md]: Change:

> Applications can be configured with much of the same options 

to:

> Applications can be configured with many of the same options 

Change:

> The difference is that an Application’s options are applied to **every instance of that Application** and an instance customisation plugin can **allow instance configuration on a per user level** and even **allow an end user to influence your instance configuration**. 

to:

>The difference is that an Application’s configuration is applied to **every instance of that Application**, whereas instance customisation plugin's **allow instance configuration on a per user level** and even **allow an end user to influence your instance configuration**.

File: docs/instance-customisation/usage.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+## Usage
+
+The following usage guide is intended for users looking to implement the above API into a Go application. Although it is possible to generate the necessary code to implement the Instance Customisation Plugin API in other languages, we will not be covering that.
```
Comment (sarahkrivan) [inline docs/instance-customisation/usage.md]: we will not be covering that here.

File: docs/instance-customisation/usage.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+The server, which you can define as
+
+```go
+type InstanceCustomisationPluginServer struct {
+   customisation.UnimplementedInstanceCustomisationPluginServer
+}
+```
+
+And the pointer receiver to implement the actual gRPC call that’s made from Scalable Pixel Streaming
```
Comment (sarahkrivan) [inline docs/instance-customisation/usage.md]: that is

File: docs/instance-customisation/what-can-i-customise.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+```json
+{
+   "user": "UserA",
+   "map": "Level_1",
+   "region": ""
+}
+```
+
+Keeping in mind that this data can represent **anything that you want**. It’s up to you what you do with this in your instance customisation plugin.
```
Comment (sarahkrivan) [inline docs/instance-customisation/what-can-i-customise.md]: It is

File: docs/instance-customisation/what-can-i-customise.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+### Runtime Options
+Runtime Options provides your instance customisation plugin with access to the configured options from your instances Application. These will be sent back to Scalable Pixel Streaming when requesting an instance, so any modifications your instance customisation plugin makes to these options will be reflected.
```
Comment (sarahkrivan) [inline docs/instance-customisation/what-can-i-customise.md]: Change:

> Runtime Options provides your instance customisation plugin with access to the configured options from your instances Application.

to:

> Runtime Options provide your instance customisation plugin access to the configured options from your instance's Application.

I think this logic checks out here for instances (aka the instance's application). Feels a little odd to have an ownership by a child process so let me revisit in the morning.


I know its obvious, but maybe change to: reflected at runtime

File: docs/instance-customisation/what-can-i-customise.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+**Volume Mounts**
+
+Mount any additional volumes to your instance, for example, unique user data.
+
+**Environment Variables**
+
+Any environment variables that you want to modify
+
+To see how you might update runtime options in your instance customisation plugin, you can refer to the examples in this repository.
```
Comment (sarahkrivan) [inline docs/instance-customisation/what-can-i-customise.md]: can we make a link here to examples

File: docs/instance-customisation/deploy.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
+## Deploying to the cloud
+
+Once you have built your instance customisation plugin it needs to be deployed to the cloud. The most common way is to deploy it to the same cluster that is running Scalable Pixel Streaming, however, you can deploy it anywhere that is publically accessible.
```
Comment (sarahkrivan) [inline docs/instance-customisation/deploy.md]: lets use publicly

**Review by sarahkrivan — CHANGES_REQUESTED** [review]

File: docs/instance-customisation/what-can-i-customise.md (hunk 7 above)
Comment (dan-tw) [reply docs/instance-customisation/what-can-i-customise.md]: Is this what you meant?

> Runtime Options provide your instance customisation plugin access to the configured options reflected at runtime.

or...

> Runtime Options provide your instance customisation plugin access to the configured options from your instance's Application reflected at runtime.

Comment (sarahkrivan) [reply docs/instance-customisation/what-can-i-customise.md]: Sorry you're right this wasn't clear enough. Lets make it:

> Runtime Options allow your instance customisation plugin to access the configurations you have set for a given Version (e.g. resolution). These configurations will be sent back to Scalable Pixel Streaming when requesting an instance, so any modifications your instance customisation plugin makes to these options will be reflected at runtime (e.g. where instance 1 has a resolution of 1920x1080 and instance 2 has a resolution of 1280x720).

**Review by sarahkrivan — APPROVED** [review]
LGTM

**Pull request merged**